neps/nep-0393.md

```diff
@@ -0,0 +1,480 @@
+NOTE: SBT smart contract (issuer) can have it's own list of blocked accounts or allowed only accounts.
+
+### Token Class
+
+SBT tokens can't be fractionized. Also, by definition there should be only one of a token per token class per user. Examples: user should not be able to receive few badges of the same class, or few proof of attendance to the same event.
+However we identify a need for having few token classes in a single contract:
+
+- badges: one contract with multiple badge class (community lead, OG...);
+- certificates: one issuer can create certificates of a different class (eg school department can create diplomas for each major and each graduation year).
+
+We also see a trend in the NFT community and demand for market places to support multi token contracts.
+
+- In Ethereum community many projects are using [ERC-1155 Multi Token Standard](https://eips.ethereum.org/EIPS/eip-1155). NFT projects are using it for fraction ownership: each token id can have many fungible fractions.
+- NEAR [NEP-245](https://github.com/near/NEPs/blob/master/neps/nep-0245.md) has elaborated similar interface for both bridge compatibility with EVM chains as well as flexibility to define different token types with different behavior in a single contract. [DevGovGigs Board](https://near.social/#/mob.near/widget/MainPage.Post.Page?accountId=devgovgigs.near&blockHeight=87938945) recently also shows growing interest to move NEP-245 adoption forward.
+- [NEP-454](https://github.com/near/NEPs/pull/454) proposes royalties support for multi token contracts.
+
+We propose that the SBT Standard will support the multi-token idea from the get go. This won't increase the complexity of the contract (in a traditional case, where one contract will only issue tokens of the single class, the `class` argument is simply ignored in the state, and in the functions it's required to be of a constant value, eg `1`) but will unify the interface.
+It's up to the smart contract design how the token classes is managed. A smart contract can expose an admin function (example: `sbt_new_class() -> ClassId`) or hard code the pre-registered classes.
+
+Finally, we require that each token ID is unique within the smart contract. This will allow us to query token only by token ID, without knowing it's class.
+
+### SBT Registry
+
+Atomicity of _soul transfer_ in current NEAR runtime is not possible if the token balance is kept separately for each SBT smart contract. We need an additional contract: the `SBT Registry`, to provide atomic transfer of all user SBTs and efficient way to ban accounts in relation to the Ban event. The registry provides a balance book of all associated SBT tokens. So, each SBT is requested by a smart contract (issuer), however it's is created in the registry.
```

Comment:
I don't think having centralised SBT Registries that actually hold these tokens is the right way to go. If this standard does want to adhere to the core ideas of NEP-171, then having contracts managing their own tokens is crucial.

Granted, it may solve the soul transfer question, however, I would suggest splitting these concerns. Having one on-chain entity only responsible for managing soul transfers, and a separate SBT contract, which knows how to communicate with this soul transfer governor.

Here is roughly what it would look like:
* You can issue a soul transfer against a governor. This would result in a soul transfer event noted in the governor. The governor can implement the various proposed validation steps.
* Each SBT contract has a list of trusted governors.
* Once a transfer is issued, you can apply the transfer on the SBT, which checks that the transfer was registered on the governor, in which case the transfer is completed

You can still burn accounts on governors, which the contracts can check themselves.

This allows the concept of recovery and soul transfer to evolve independently of the SBT Contracts themselves.
In your proposed solution, your SBTs are locked to the registry forever. Thus if a registry is vulnerable, all of those tokens are lost.

In the solution above, SBTs can pick which governors they support, thus not having to keep all state in sync in multiple registries. Instead, there is a single point of truth for the SBT, and multiple potential approaches to soul transfer and recovery. I'm sure you can see how this is more inline with the concept of SBTs.

Reply:
Registry could be fully permissionless without any admin feature. Again, we don't define in the standard how the registry governance should work. Hopefully there will be good market for it (as for example DEXes).

> Yes, except here one NFT collection is vulnerable. Not anyone who opted to use the registry.

Same for marketplaces, aggregators, escrows, exchanges....

Simple registry implementation will go through many reviews and we can do audits. Could be a template.

Reply:
So what exactly happens if you do a soul transfer on one registry and not another? How are registries kept in sync with each other, for changes that occur on the registry?

> Simple registry implementation will go through many reviews and we can do audits

Yes, but with a governor you could just have an audited version deploy that others can use. You are decoupling SBT behavour and issuance and recovery/transfer.

Reply:
> https://github.com/near/NEPs/pull/393/#discussion_r1163938750

It depends on you. We can't restrict issuers to issue only one "copy" of SBT, and we can't restrict registries to be unique. In a same way, we can't restrict that NFT collections won't be copied, or issued twice (on the same chain or multiple chains).  
Ultimately, the credible ones will be used.

Reply:
Yes, except that NFTs are designed to be a single source of truth. So a different contract is a different NFT. And you can clearly tell with branding and marketing which one is the real one, and which one is the copy.

The design here does not have a single source of truth. The same SBT on different registries are supposed to be the same. And there is no method of arbitration of which are supposed to be the "real" ones.

I feel like this is a fundamental flaw in this design, and it would be a mistake to go ahead as it is.